icu4c/source/common/mlbe.cpp

```diff
@@ -128,135 +135,127 @@ int32_t MlBreakEngine::divideUpRange(UText *inText, int32_t rangeStart, int32_t
-    }
-    if (elementList[1] != INVALID && elementList[2] != INVALID && elementList[3] != INVALID) {
-        score += (2 * fModel.geti(feature.setTo(u"TW2:", 4)
-                                      .append(elementList[1])
-                                      .append(elementList[2])
-                                      .append(elementList[3])));
+    for (int i = 0; i < 6; i++) {
+        // UW1 ~ UW5
+        start = startIdx + i;
+        if (indexList[start] != -1) {
+            end = (indexList[start + 1] != -1) ? indexList[start + 1] : numCodeUnits;
+            score +=
+                fModel[i].geti(inString.tempSubString(indexList[start], end - indexList[start]));
+        }
     }
-    if (elementList[2] != INVALID && elementList[3] != INVALID && elementList[4] != INVALID) {
-        score += (2 * fModel.geti(feature.setTo(u"TW3:", 4)
-                                      .append(elementList[2])
-                                      .append(elementList[3])
-                                      .append(elementList[4])));
+    for (int i = 0; i < 3; i++) {
+        // BW1 ~ BW3
+        start = startIdx + i + 1;
+        if (indexList[start] != -1 && indexList[start + 1] != -1) {
+            end = (indexList[start + 2] != -1) ? indexList[start + 2] : numCodeUnits;
+            score += fModel[i + 6].geti(
+                inString.tempSubString(indexList[start], end - indexList[start]));
+        }
     }
-    if (elementList[3] != INVALID && elementList[4] != INVALID && elementList[5] != INVALID) {
-        score += (2 * fModel.geti(feature.setTo(u"TW4:", 4)
-                                      .append(elementList[3])
-                                      .append(elementList[4])
-                                      .append(elementList[5])));
+    for (int i = 0; i < 4; i++) {
+        // TW1 ~ TW4
+        start = startIdx + i;
+        if (indexList[start] != -1 && indexList[start + 1] != -1 && indexList[start + 2] != -1) {
+            end = (indexList[start + 3] != -1) ? indexList[start + 3] : numCodeUnits;
+            score += fModel[i + 9].geti(
+                inString.tempSubString(indexList[start], end - indexList[start]));
+        }
     }
+
     if (score > 0) {
-        boundary.addElement(index, status);
+        boundary.addElement(breakpointIdx, status);
         numBreaks++;
     }
+    return numBreaks;
 }
 
-int32_t MlBreakEngine::initElementList(const UnicodeString &inString, UChar32* elementList,
-                                         UErrorCode &status) const {
+int32_t MlBreakEngine::initIndexList(const UnicodeString &inString, int32_t *indexList,
+                                     UErrorCode &status) const {
     if (U_FAILURE(status)) {
         return 0;
     }
     int32_t index = 0;
     int32_t length = inString.countChar32();
-    UChar32 w1, w2, w3, w4, w5, w6;
-    w1 = w2 = w3 = w4 = w5 = w6 = INVALID;
+    for (int i = 0; i < length + 4; i++) {
```

Comment:
is thsi correct?
Let's say you have lenght == 1. your allocated indexList is size of 5
after this loop, you will have
-1, -1, -1, -1, -1 
and after after the function return, you will have
-1, -1, 0, -1, -1
is that correct?

Reply:
Yes